README.md
# TermTetris
A Tetris game in character terminal evironment.


CHANGELOG
0.9.16 Current release
  * have a taste of namedtuple and data class.
0.9.13
  * fix 'has no attribute keyparam' after upgrade ltermio to 0.2.7.
0.9.12
  * fix the 'getkey() is not defined' while press esc to pause game.
0.9.11
  * Modify code to adapt the interface changing of the termkey module.
0.9.10
  * Rename the game and source files to resolve conflicts.
0.9.9
  + Almost a stable version with complete functionality.


src/panel.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""Two types of game panels are defined in the module: ActivePanel and
MessagePanel.

The ActivePanel is the main activity area of the game, player plays the game
on it. The MessagePanel is a panel to display message, such as game score,
next tetro, greeting and help text, etc.
"""

import time
import functools
from wcwidth import wcswidth

from cursor import putmsg, downward_seq, rect_border_seq
import color256 as co
from unicon import UnicodeIcon as UIcon

from tetro import Tetro


_EDGE_SQUARE = UIcon.BROWN_SQUARE
_BLANK = UIcon.WHITE_SQUARE


def _color(func):
    # A decorator to set and restore the color setting.
    @functools.wraps(func)
    def wrapper(self, *args, **kwargs):
        if self.color_scheme:
            co.set_color(self.color_scheme[0], self.color_scheme[1])
        try:
            return func(self, *args, **kwargs)
        finally:
            if self.color_scheme:
                co.set_color(self.color_scheme[2], self.color_scheme[3])
    return wrapper


class ActivePanel:  # pylint: disable=too-many-instance-attributes
    """The panel displays the main activity area of the game, player plays
    the game on it.
    """
    def __init__(
        self,
        o_row: int,
        o_col: int,
        width: int = 10,
        height: int = 20,
        *,
        blank: str = _BLANK,
    ):
        self.o_row = o_row
        self.o_col = o_col
        self.width = width
        self.height = height
        self.blank = blank
        self.color_scheme = None
        self.image = [[blank for _ in range(self.width)]
                      for _ in range(self.height)]
        self.t_row: int
        self.t_col: int
        self.tetro = None
        self._blank_row = blank * width


    @_color
    def show(self):
        """Shows the panel on the screen.
        """
        putmsg(self.o_row - 1, self.o_col - 2,
               rect_border_seq(self.width + 2, self.height + 2, _EDGE_SQUARE))
        self.refresh()


    @_color
    def refresh(self):
        """Refreshs displaying of the activity area, excludes border frame.
        """
        putmsg(self.o_row, self.o_col,
               ''.join([
                   downward_seq(''.join(self.image[row]), self.width * 2)
                   for row in range(self.height)])
               )
        self._show_tetro()


    def clear(self):
        """Clears all the squares(including tetro) out of the panel.
        """
        for row in range(self.height):
            for col in range(self.width):
                self.image[row][col] = self.blank
        self.tetro = None
        self.refresh()


    def _show_tetro(self, *, hide: bool = False):
        if self.tetro:
            self.tetro.show(
                    self.o_row + self.t_row, self.o_col + self.t_col * 2,
                    self.blank if hide else self.tetro.square)


    def tetro_pos(self, row: int, col: int):
        """Sets the coordinate position of the tetro.
        """
        self.t_row = row
        self.t_col = col


    @_color
    def refresh_tetro(
        self,
        tetro: Tetro,
        *,
        row: int | None = None,
        col: int | None = None,
    ):
        """Replaces the current tetro with a new tetro(could be the same
        one) and displays it.

        If a coordinate(row, col) is specified, the new tetro will be
        displayed at there.
        """
        self._show_tetro(hide=True)
        if row is not None:
            self.t_row = row
        if col is not None:
            self.t_col = col
        self.tetro = tetro
        self._show_tetro()


    def merge_tetro(self):
        """Merges the current tetro into background grid.
        """
        for i in range(6):
            if (self.tetro.bits >> i) & 0x01:
                row = self.t_row + i // self.tetro.width
                col = self.t_col + i % self.tetro.width
                self.image[row][col] = self.tetro.square
        self.tetro = None


    def _show_rows(self, row_list: list[int], *, hide: bool = False):
        for row in row_list:
            putmsg(self.o_row + row, self.o_col,
                   self._blank_row if hide else ''.join(self.image[row]))


    @_color
    def blink_rows(self, row_list: list[int], times: int, interval: float):
        """Blinks the specified rows 'times' with 'interval' time.
        """
        for _ in range(times):
            self._show_rows(row_list, hide=True)
            time.sleep(interval)
            self._show_rows(row_list)
            time.sleep(interval)


    def remove_rows(self, row_list: list[int]):
        """Clears the speicified rows.
        """
        for row in row_list:
            self.image.pop(row)
            self.image.insert(0, [self.blank for _ in range(self.width)])
        self.refresh()


_MARGIN = 1
_SEPARATOR = UIcon.SMALL_WHITE_STAR * 2

class MessagePanel(ActivePanel):
    """A panel to display message. It is derived from the ActivePanel.
    """
    def __init__(
        self,
        o_row: int,
        o_col: int,
        width: int = 10,
        height: int = 20
    ):
        super().__init__(o_row, o_col, width, height, blank='  ')
        self._row_cursor = 0


    LEFT = 0
    RIGHT = 1
    CENTER = 2

    @_color
    def put_text(
        self,
        text: str,
        *,
        align: int | None = None,
        row: int | None = None,
    ):
        """Outputs text on the panel.

        Args:
            text: The text to output. It can use '\n' to wrap around, the
                row naturally increments up to down.
            align: Value of MessagePanel.CENTER aligns text to the center,
                MessagePanel.RIGHT aligns text to the right. Default left
                alignment.
            row: If specified, the output starts from the 'row'.
        """
        if row is not None:
            self._row_cursor = row
        lines = text.split(sep='\n')
        for txt in lines:
            if align == self.CENTER:
                offset = (self.width * 2 - wcswidth(txt)) // 2
            elif align == self.RIGHT:
                offset = self.width * 2 - wcswidth(txt) - _MARGIN
            else:
                offset = _MARGIN
            putmsg(self.o_row + self._row_cursor,
                   self.o_col + offset,
                   txt[:(self.width - _MARGIN) * 2])
            self._row_cursor += 1


    @_color
    def add_separator(self, *, row: int | None = None):
        """Adds a separator line at current row or the 'row' if specified.
        """
        if row is not None:
            self._row_cursor = row
        putmsg(self.o_row + self._row_cursor, self.o_col + _MARGIN,
               f'{_SEPARATOR * (self.width - 1)} ')
        self._row_cursor += 1


src/tetro.py
# -*- coding: utf-8 -*-

"""Pre-defines the constant instances of all types and shapes of the tetrominos.
"""

import random

from cursor import putmsg, v_composing
from unicon import UnicodeIcon as UIcon


_P1 = ':{0}\x1b2hj'
_P01 = '2l:{0}\x1b4hj'
_P10 = _P1
_P11 = '2:{0}\x1b4hj'
_P001 = '4l:{0}\x1b6hj'
_P010 = _P01
_P011 = '2l2:{0}\x1b6hj'
_P100 = _P1
_P101 = ':{0}\x1b2l:{0}\x1b6hj'
_P110 = _P11
_P111 = '3:{0}\x1b6hj'
_P1111 = '4:{0}'


_SHAPE_I0 = v_composing(_P1 * 4)
_SHAPE_I1 = v_composing(_P1111)
_SHAPE_J0 = v_composing(f'{_P01 * 2}{_P11}')
_SHAPE_J1 = v_composing(f'{_P111}{_P001}')
_SHAPE_J2 = v_composing(f'{_P11}{_P10 * 2}')
_SHAPE_J3 = v_composing(f'{_P100}{_P111}')
_SHAPE_L0 = v_composing(f'{_P10 * 2}{_P11}')
_SHAPE_L1 = v_composing(f'{_P001}{_P111}')
_SHAPE_L2 = v_composing(f'{_P11}{_P01 * 2}')
_SHAPE_L3 = v_composing(f'{_P111}{_P100}')
_SHAPE_O0 = v_composing(f'{_P11 * 2}')
_SHAPE_S0 = v_composing(f'{_P011}{_P110}')
_SHAPE_S1 = v_composing(f'{_P10}{_P11}{_P01}')
_SHAPE_T0 = v_composing(f'{_P111}{_P010}')
_SHAPE_T1 = v_composing(f'{_P10}{_P11}{_P10}')
_SHAPE_T2 = v_composing(f'{_P010}{_P111}')
_SHAPE_T3 = v_composing(f'{_P01}{_P11}{_P01}')
_SHAPE_Z0 = v_composing(f'{_P110}{_P011}')
_SHAPE_Z1 = v_composing(f'{_P01}{_P11}{_P10}')


class Tetro:
    """Data structure description of the tetrominos.

    All types and shapes of tetrominos are pre-defined as constants and
    their instances are stored in a pool, in order of alphabetically
    sorted.
    """
    SQUARE_I = UIcon.GREEN_SQUARE
    SQUARE_J = UIcon.BLUE_SQUARE
    SQUARE_L = UIcon.ORANGE_SQUARE
    SQUARE_O = UIcon.YELLOW_SQUARE
    SQUARE_S = UIcon.GREEN_SQUARE
    SQUARE_T = UIcon.MAGENTA_SQUARE
    SQUARE_Z = UIcon.RED_SQUARE


    def __init__(  # pylint: disable=too-many-arguments
        self,
        height: int,
        width: int,
        bits: int,
        square: str,
        shape: str,
    ):
        self.height = height
        self.width = width
        self.bits = bits
        self.square = square
        self.shape = shape


    def show(self, row: int, col: int, square: str):
        """Shows tetro on the specified coordinate by using the 'square' as
        the composing unit.
        """
        putmsg(row, col, self.shape.format(square))


    def rotate(self):
        """Returns a new instance of rotating shape of the tetro.
        """
        i = TETRO_POOL.index(self)
        return TETRO_POOL[(i & 0xfc) | ((i + 1) & 0x03)]


    @staticmethod
    def choice():
        """Returns a random choiced tetro instance from the pool of tetrominos.
        """
        return random.choice(TETRO_POOL)


TETRO_I_0 = Tetro(4, 1, 0x0f, Tetro.SQUARE_I, _SHAPE_I0)
TETRO_I_1 = Tetro(1, 4, 0x0f, Tetro.SQUARE_I, _SHAPE_I1)
TETRO_I_2 = TETRO_I_0
TETRO_I_3 = TETRO_I_1

TETRO_J_0 = Tetro(3, 2, 0x3a, Tetro.SQUARE_J, _SHAPE_J0)
TETRO_J_1 = Tetro(2, 3, 0x27, Tetro.SQUARE_J, _SHAPE_J1)
TETRO_J_2 = Tetro(3, 2, 0x17, Tetro.SQUARE_J, _SHAPE_J2)
TETRO_J_3 = Tetro(2, 3, 0x39, Tetro.SQUARE_J, _SHAPE_J3)

TETRO_L_0 = Tetro(3, 2, 0x35, Tetro.SQUARE_L, _SHAPE_L0)
TETRO_L_1 = Tetro(2, 3, 0x3c, Tetro.SQUARE_L, _SHAPE_L1)
TETRO_L_2 = Tetro(3, 2, 0x2b, Tetro.SQUARE_L, _SHAPE_L2)
TETRO_L_3 = Tetro(2, 3, 0x0f, Tetro.SQUARE_L, _SHAPE_L3)

TETRO_O_0 = Tetro(2, 2, 0x0f, Tetro.SQUARE_O, _SHAPE_O0)
TETRO_O_1 = TETRO_O_0
TETRO_O_2 = TETRO_O_0
TETRO_O_3 = TETRO_O_0

TETRO_S_0 = Tetro(2, 3, 0x1e, Tetro.SQUARE_S, _SHAPE_S0)
TETRO_S_1 = Tetro(3, 2, 0x2d, Tetro.SQUARE_S, _SHAPE_S1)
TETRO_S_2 = TETRO_S_0
TETRO_S_3 = TETRO_S_1

TETRO_T_0 = Tetro(2, 3, 0x17, Tetro.SQUARE_T, _SHAPE_T0)
TETRO_T_1 = Tetro(3, 2, 0x1d, Tetro.SQUARE_T, _SHAPE_T1)
TETRO_T_2 = Tetro(2, 3, 0x3a, Tetro.SQUARE_T, _SHAPE_T2)
TETRO_T_3 = Tetro(3, 2, 0x2e, Tetro.SQUARE_T, _SHAPE_T3)

TETRO_Z_0 = Tetro(2, 3, 0x33, Tetro.SQUARE_Z, _SHAPE_Z0)
TETRO_Z_1 = Tetro(3, 2, 0x1e, Tetro.SQUARE_Z, _SHAPE_Z1)
TETRO_Z_2 = TETRO_Z_0
TETRO_Z_3 = TETRO_Z_1

TETRO_POOL = (
    TETRO_I_0, TETRO_I_1, TETRO_I_2, TETRO_I_3,
    TETRO_J_0, TETRO_J_1, TETRO_J_2, TETRO_J_3,
    TETRO_L_0, TETRO_L_1, TETRO_L_2, TETRO_L_3,
    TETRO_O_0, TETRO_O_1, TETRO_O_2, TETRO_O_3,
    TETRO_S_0, TETRO_S_1, TETRO_S_2, TETRO_S_3,
    TETRO_T_0, TETRO_T_1, TETRO_T_2, TETRO_T_3,
    TETRO_Z_0, TETRO_Z_1, TETRO_Z_2, TETRO_Z_3,
)


src/backend.py
# -*- coding: utf-8 -*-

"""Backend service for the Tetris game.
"""

import functools

from tetro import Tetro


def _check_tetro(func):
    # A decorator to check if the current tetro is valid.
    @functools.wraps(func)
    def wrapper(self, *args, **kwargs):
        if self.tetro is None:
            return None
        return func(self, *args, **kwargs)
    return wrapper


class TetrisBackend:
    """Backend service for the Tetris game.

    The backend guides the game by maintaining a internal bitmap grid and
    the tetro's info, it determines which operations of frontend are valid.
    Almost all of its public methods have an uniform returning format of
    tuple on success, except print_grid() which is only for debugging.

    The result tuple is in form of (tetro, row, col, elim_rows), the first
    three parameters indicates what and where the tetro should be shown, the
    last one contains row numbers which have been eliminated if there were.
    """
    def __init__(self, width: int = 10, height: int = 20):
        self.width = width
        self.height = height
        self.tetro: Tetro | None = None
        self.next_tetro: Tetro = Tetro.choice()
        self.t_row: int
        self.t_col: int
        self.grid: list[int] = [0] * height


    def kick_off(self) -> tuple[Tetro, int, int, list[int] | None] | None:
        """Start a new game.
        """
        for i in range(self.height):
            self.grid[i] = 0
        return self._issue_next(None)


    def _issue_next(self, elim_rows: list[int] | None):
        self.tetro = self.next_tetro
        self.t_row = 0
        self.t_col = (self.width - self.tetro.width) // 2
        if self._is_overlapped(self.t_row, self.t_col):
            self.tetro = None
            return None
        self.next_tetro = Tetro.choice()
        return self.tetro, self.t_row, self.t_col, elim_rows


    def _is_overlapped(self, row: int, col: int) -> bool:
        if not (0 <= col <= self.width - self.tetro.width and
                row + self.tetro.height <= self.height):
            return True
        bits = 0
        mask = (1 << self.tetro.width) - 1
        for i in range(self.tetro.height):
            bits |= (((self.grid[row + i] >> col) & mask)
                     << (i * self.tetro.width))
        return bool(bits & self.tetro.bits)


    def _merge(self):
        mask = (1 << self.tetro.width) - 1
        for i in range(self.tetro.height):
            self.grid[self.t_row + i] |= (
                    ((self.tetro.bits >> (i * self.tetro.width)) & mask)
                    << self.t_col)
        self.tetro = None


    def _check_elims(self) -> list[int]:
        result = []
        for i, row in enumerate(self.grid):
            if row == (1 << self.width) - 1:
                result.append(i)
                self.grid.pop(i)
                self.grid.insert(0, 0)
        return result


    @_check_tetro
    def _hori_move(self, offset: int):
        if self._is_overlapped(self.t_row, self.t_col + offset):
            return None
        self.t_col += offset
        return self.tetro, self.t_row, self.t_col, None


    def move_left(self) -> tuple[Tetro, int, int, None] | None:
        """Moves the tetro to the left one column.

        Returns:
            The tetro's info in a tuple after sucessful move, None if the
            left column is out of the grid bound.
        """
        return self._hori_move(-1)


    def move_right(self) -> tuple[Tetro, int, int, None] | None:
        """Moves the tetro to the right one column.

        Returns:
            The tetro's info in a tuple after sucessful move, None if the
            right column is out of the grid bound.
        """
        return self._hori_move(1)


    @_check_tetro
    def move_down(self) -> tuple[Tetro, int, int, list[int] | None] | None:
        """Moves the tetro down one row.

        If there is no more space to move down, it triggers the current
        tetro to be merged into the grid and elimination checking, then
        returns the eliminated row numbers if there were, and a new tetro
        which be issued at the top of the grid.
        """
        if self._is_overlapped(self.t_row + 1, self.t_col):
            self._merge()
            elim_rows = self._check_elims()
            return self._issue_next(elim_rows)
        self.t_row += 1
        return self.tetro, self.t_row, self.t_col, None


    @_check_tetro
    def fall_down(self) -> tuple[Tetro, int, int, None] | None:
        """Makes the current tetro to fall down to ground.
        """
        i = 1
        while not self._is_overlapped(self.t_row + i, self.t_col):
            i += 1
        self.t_row += i - 1
        return self.tetro, self.t_row, self.t_col, None


    @_check_tetro
    def rotate(self) -> tuple[Tetro, int, int, None] | None:
        """Rotates the current tetro clockwise.
        """
        old = self.tetro
        self.tetro = old.rotate()
        # Calculates the appropriate coordinate after rotation
        col = (old.width - self.tetro.width) // 2
        col += self.t_col + 1 if col < 0 else self.t_col
        col = min(max(0, col), self.width - self.tetro.width)

        if self._is_overlapped(self.t_row, col):
            self.tetro = old
            return None
        self.t_col = col
        return self.tetro, self.t_row, self.t_col, None


    def print_grid(self):
        """Prints bitmap of the grid, only for debugging.
        """
        print('')
        for i, row in enumerate(self.grid):
            bits = bin(row)[:1:-1]
            print(f'{i:02d}: {bits}{"0" * (self.width - len(bits))}')


termtris/termtris.py
# -*- coding: utf-8 -*-

#   Copyright 2023 Brooks Su
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

"""The controller and application entry point of the Termtris game.
"""

import os
import sys

import ltermio
from ltermio import getkey, Key, Color

from .tt_tetro import Tetro
from .tt_backend import TetrisBackend, TetroUpdate
from .tt_panel import ActivePanel, MessagePanel


_SCORE_TABLE = (10, 100, 200, 400, 800)
_LEVEL_TABLE = (1000, 2000, 4000, 8000, 16000, 32000, 64000, 80_000_000)


class Termtris():
    # pylint: disable=too-few-public-methods
    # pylint: disable=too-many-instance-attributes
    """The controller of the Termtris game.

    The class have only one public method run() which reads keys from input
    and gets appropriate methods of the backend and the panels to process
    the keyboard events.
    """
    def __init__(self, o_row: int, o_col: int, width: int, height: int):
        self.act_panel = ActivePanel(o_row, o_col, width, height)
        self.msg_panel = MessagePanel(
                o_row, o_col + (width + 1) * 2, width, height)
        self.backend = TetrisBackend(width, height)

        self.score: int = 0
        self.highest: int = 0
        self.speed: int
        self.tick: int = 0
        self.stat_sect_row = height - 5


    def _init_msg_panel(self):
        self.msg_panel.put_text('Termtris', align=MessagePanel.CENTER)
        self.msg_panel.add_separator()
        self.msg_panel.put_text(
                'Right:  Move Right\n'
                'Left:   Move Left\n'
                'Up:     Rotate\n'
                'Down:   Speed up Fall\n'
                'Space:  Fall to Ground\n'
                'Enter:  New Game\n'
                'Esc:    Pause Game\n'
                'Ctrl-X: Exit Game')
        self.msg_panel.add_separator()

        self.msg_panel.add_separator(row=self.stat_sect_row)
        self.msg_panel.tetro_pos(self.stat_sect_row + 1, 8)
        self._renew_game_stat()


    def _renew_game_stat(self):
        for i, level, in enumerate(_LEVEL_TABLE):
            if self.score < level:
                self.msg_panel.put_text(
                        'Next Tetro:\n'
                        f'Level: {i + 1}\n'
                        f'Score: {self.score:<6d}\n'
                        f'Highest: {self.highest:<6d}',
                        row=self.stat_sect_row + 1)
                self.msg_panel.refresh_tetro(self.backend.next_tetro)
                self.speed = len(_LEVEL_TABLE) - i
                break


    def _update_score(self, row_num: int):
        self.score += _SCORE_TABLE[row_num]
        if self.score > self.highest:
            self.highest = self.score
        self._renew_game_stat()


    def _new_game(self) -> None | tuple[Tetro, int, int, list[int] | None]:
        self.act_panel.clear()
        res = self.backend.kick_off()
        self.score = 0
        self._update_score(0)
        return res


    def _idle_fall(self) -> None | tuple[Tetro, int, int, list[int] | None]:
        self.tick = (self.tick + 1) % self.speed
        return None if self.tick else self.backend.move_down()


    def _update_tetro(
        self,
        tetro: Tetro,
        row: int,
        col: int,
        elims: list[int],
    ):
        if elims is not None:  # Current tetro has done
            self.act_panel.merge_tetro()
            if elims:
                self.act_panel.blink_rows(elims, 3, 0.1)
                self.act_panel.remove_rows(elims)
            self._update_score(len(elims))
        self.act_panel.refresh_tetro(tetro, row=row, col=col)


    def run(self):
        """Continuously reads keys from input and gets appropriate functions
        to handle key events, then according to the event results, refreshs
        the displaying of the active panel.
        """
        key_funcs = {
            Key.NONE: self._idle_fall,
            Key.ENTER: self._new_game,
            Key.ESC: lambda: not ltermio.getch(),
            Key.DOWN: self.backend.move_down,
            Key.SPACE: self.backend.fall_down,
            Key.UP: self.backend.rotate,
            Key.RIGHT: self.backend.move_right,
            Key.LEFT: self.backend.move_left,
            Key.CONTROL_D: self.backend.print_grid,
        }
        self.act_panel.show()
        self.msg_panel.show()
        self._init_msg_panel()

        key = getkey()
        while key != Key.CONTROL_X:
            update = key_funcs.get(key, self._idle_fall)()
            if update:
                self._update_tetro(*update)
            key = getkey(1)


@ltermio.appentry
def main():
    """Entry point of the Termtris game.

    Detects terminal environment and makes initial arguments of the game.
    """
    width, height = ((12, 20) if len(sys.argv) < 3 else
                     (int(sys.argv[1]), int(sys.argv[2])))
    if width < 12 or height < 20:
        raise ValueError('The width must be no less than 12 and the height 20')

    scr_width, scr_height = os.get_terminal_size()
    o_row = min((scr_height - (height + 2)) // 2 + 1, 5)
    o_col = (scr_width - (width * 2 + 3) * 2) // 2 + 1
    if o_row <= 0 or o_col <= 0:
        raise EnvironmentError('Screen too small to fit game')

    ltermio.set_color(Color.DEEP_KHAKI, Color.COFFEE)
    Termtris(o_row, o_col, width, height).run()
    ltermio.reset_color()


src/tetris.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""The controller and application entry point of the Tetris game.
"""

import os
import sys

import cursor
import color256 as co
from color256 import Color
from termkey import getkey, Key

from tetro import Tetro
from backend import TetrisBackend
from panel import ActivePanel, MessagePanel


_SCORE_TABLE = (10, 100, 200, 400, 800)
_LEVEL_TABLE = (1000, 2000, 4000, 8000, 16000, 32000, 64000, 80_000_000)


class Tetris():
    # pylint: disable=too-few-public-methods
    # pylint: disable=too-many-instance-attributes
    """The controller of the Tetris game.

    The class have only one public method run() which reads keys from input
    and gets appropriate methods of the backend and the panels to process the
    keyboard events.
    """
    def __init__(self, o_row: int, o_col: int, width: int, height: int):
        self.act_panel = ActivePanel(o_row, o_col, width, height)
        self.msg_panel = MessagePanel(
                o_row, o_col + (width + 1) * 2, width, height)
        self.backend = TetrisBackend(width, height)

        self.score: int = 0
        self.highest: int = 0
        self.speed: int
        self.tick: int = 0
        self.stat_sect_row = height - 5


    def _init_msg_panel(self):
        self.msg_panel.put_text('TermTetris', align=MessagePanel.CENTER)
        self.msg_panel.add_separator()
        self.msg_panel.put_text(
                'Right:  Move Right\n'
                'Left:   Move Left\n'
                'Up:     Rotate\n'
                'Down:   Speed up Fall\n'
                'Space:  Fall to Ground\n'
                'Enter:  New Game\n'
                'Esc:    Pause Game\n'
                'Ctrl-X: Exit Game')
        self.msg_panel.add_separator()

        self.msg_panel.add_separator(row=self.stat_sect_row)
        self.msg_panel.tetro_pos(self.stat_sect_row + 1, 8)
        self._renew_game_stat()


    def _renew_game_stat(self):
        for i, level, in enumerate(_LEVEL_TABLE):
            if self.score < level:
                self.msg_panel.put_text(
                        'Next Tetro:\n'
                        f'Level: {i + 1}\n'
                        f'Score: {self.score:<6d}\n'
                        f'Highest: {self.highest:<6d}',
                        row=self.stat_sect_row + 1)
                self.msg_panel.refresh_tetro(self.backend.next_tetro)
                self.speed = len(_LEVEL_TABLE) - i
                break


    def _update_score(self, row_num: int):
        self.score += _SCORE_TABLE[row_num]
        if self.score > self.highest:
            self.highest = self.score
        self._renew_game_stat()


    def _new_game(self) -> None | tuple[Tetro, int, int, list[int] | None]:
        self.act_panel.clear()
        res = self.backend.kick_off()
        self.score = 0
        self._update_score(0)
        return res


    def _idle_fall(self) -> None | tuple[Tetro, int, int, list[int] | None]:
        self.tick = (self.tick + 1) % self.speed
        return None if self.tick else self.backend.move_down()


    def _act_response(
        self,
        tetro: Tetro,
        row: int,
        col: int,
        elim_rows: list[int] | None,
    ):
        if elim_rows is not None:  # Current tetro has done
            self.act_panel.merge_tetro()
            if elim_rows:
                self.act_panel.blink_rows(elim_rows, 3, 0.1)
                self.act_panel.remove_rows(elim_rows)
            self._update_score(len(elim_rows))
        self.act_panel.refresh_tetro(tetro, row=row, col=col)


    def run(self):
        """Continuously reads keys from input and gets appropriate functions
        to process these key events. In idle time, makes the current tetro to
        fall down by count ticks.
        """
        key_funcs = {
            Key.NONE: self._idle_fall,
            Key.ENTER: self._new_game,
            Key.ESC: lambda: not getkey(),
            Key.DOWN: self.backend.move_down,
            Key.SPACE: self.backend.fall_down,
            Key.UP: self.backend.rotate,
            Key.RIGHT: self.backend.move_right,
            Key.LEFT: self.backend.move_left,
            Key.CONTROL_D: self.backend.print_grid,
        }
        self.act_panel.show()
        self.msg_panel.show()
        self._init_msg_panel()

        key = getkey()
        while key != Key.CONTROL_X:
            res = key_funcs.get(key, self._idle_fall)()
            if res:  # (tetro, row, col, elim_rows)
                self._act_response(res[0], res[1], res[2], res[3])
            key = getkey(1)


def main():
    """Entry point of the Tetris game.

    Detects terminal environment and makes initial arguments of the game.
    """
    width, height = ((12, 20) if len(sys.argv) < 3 else
                     (int(sys.argv[1]), int(sys.argv[2])))
    if width < 12 or height < 20:
        raise ValueError('The width must be no less than 12 and the height 20')

    scr_width, scr_height = os.get_terminal_size()
    o_row = min((scr_height - (height + 2)) // 2 + 1, 5)
    o_col = (scr_width - (width * 2 + 3) * 2) // 2 + 1
    if o_row <= 0 or o_col <= 0:
        raise EnvironmentError('Screen too small to fit game')

    cursor.switch_screen()
    cursor.clear_screen()
    cursor.hide_cursor()
    co.set_color(Color.DEEP_KHAKI, Color.COFFEE)

    try:
        Tetris(o_row, o_col, width, height).run()
    finally:
        co.reset_color()
        cursor.show_cursor()
        cursor.restore_screen()


if __name__ == '__main__':
    main()
